gpjax/gps.py

```diff
@@ -611,12 +611,17 @@ class NonConjugatePosterior(AbstractPosterior):
     from, or optimise an approximation to, the posterior distribution.
     """
 
-    latent: Float[Array, "N 1"] = param_field(None)
+    _latent: Float[Array, "N 1"] = param_field(None)
     key: KeyArray = static_field(PRNGKey(42))
 
-    def __post_init__(self):
-        if self.latent is None:
-            self.latent = normal(self.key, shape=(self.likelihood.num_datapoints, 1))
+    def get_latent(self, num_datapoints: int) -> Float[Array, "N 1"]:
+        if self._latent is None:
+            self._latent = normal(self.key, shape=(num_datapoints, 1))
+        elif self._latent.shape[0] != num_datapoints:
+            raise ValueError(
+                "The number of datapoints supplied is not equal to the number of datapoints used in the existing latent function."
+            )
+        return self._latent
```

Comment:
Careful. This look like a stateful method to me - which means not compatible with JAX tracers.

Also, I am surprised that `self._latent = normal...` assignment does not throw an error, as the `Module` should be immutable?

Reply:
Thanks for the review @daniel-dodd. For your latter point - it may well throw an error, I haven't actually tested it yet, was just looking for some feedback on the approach before refactoring all tests etc. 

For the former point, I see this could be a problem, and the side-effects didn't feel ideal. Do you think a better approach would be to simply do something like `return normal(self.key, shape=(num_datapoints, 1))` inside `get_latent()` function, and stop having a `latent` attribute? However, I guess this might be not ideal, as if the number of datapoints changes, the latent function values will change (I'm not sure if this lack of "consistency" is an issue though as I'm not too familiar with this NonConjugatePosterior stuff).

Reply:
I agree with @daniel-dodd that we likely have impure functions here which is going to break tracers. Further, the issue with not having the latent function's values stored as a field of the model is likely to make it tricky, if not impossible, to differentiate against the latent function's values.

Reply:
Maybe there's something more clever than I'm capable of conjuring up though that would solve this.

Reply:
Yeah, this is a tricky one. 

From my experience using GPJax for experimental design, specifying data point numbers can be frustrating. I typically worked with sparse GPs to independently set the model dimensions (via inducing points) of the datapoints dimension and could just pass e.g., approaches like `likelihood(num_datapoints=123)`. While not universally perfect, it resolved issues and aligned with scalability needs.

To initialise the NonConjugatePosterior model state, knowing the abstract data shape is crucial. This determines initial values for the latent variable, refining during training. This process parallels flax's `.init(key, dummy_x)` concept. The code structure currently demands awareness of the "`dummy_x`" shape's placement. Alternately, in GPFlow, posterior models are defined by upfront data provision: `model = gpflow.models.GPR(data, ...)` or `model = gpflow.models.GPMC(data,...)`. Is this data-forward approach more convenient than passing data to an objective function and abstract form (in terms of the datapoint number specification) to the likelihood?"